Reduce empty compaction tasks

bhalevy opened the pull request:
This short series prevents the creation of compaction tasks when we know in advance that they have nothing to do.
This is possible in the clean path by:
- improve the detection of candidates for cleanup by skipping sstables that require cleanup but are already being compacted
- checking that list of sstables selected for cleanup isn't empty before creating the cleanup task

For upgrade sstables, and generally when rewriting all sstable: launch the task only if the list off candidate sstables isn't empty.

For regular compaction, when triggered via `submit` that's called en-mass when the `compaction_reenabler` is destroyed upon exiting `run_with_compaction_disabled`, calculate the candidate sstables list in `compaction_reenabler` dtor, while compaction is disabled and pass it down to submit, so it can check if it's empty before launching (in the background) the regular compaction task executor.

Refs scylladb/scylladb#15673
Refs scylladb/scylladb#16694
Fixes scylladb/scylladb#16803
Fixes scylladb/scylladb#16804

bhalevy commented:
Cc @denesb @Deexie

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -1267,19 +1271,23 @@ class regular_compaction_task_executor : public compaction_task_executor, public
+    if (candidates_opt && candidates_opt->empty()) {
+        return;
+    }
+
```
this is candidates for the compaction strategy.

but who evaluates the need for a task is the compaction strategy.

could happen that a strategy is satisfied but candidates_opt will not be empty, and we'd fall into the same problem.

Additionally, we went a long way to allow compaction strategy to receive either run or raw sstables, so sstables are today only fed into strategy by control. That was to avoid expensive cpu cycles to convert sstables into runs, in ICS.

What I believe is the proper way to tackle this problem is to decouple compaction_manager task from task_manager task (for regular compaction). We cannot avoid creating a compaction_manager task (as we have to wait for an async task on stop()), but we can avoid creating a task_manager one. Every time compaction_manager task decides to run a regular compaction, then we create a task_manager task for it.

bhalevy replied:
Why can't we avoid creating a compaction_manager task?
Say the compaction strategy would expose a heuristic function that can tell if definitely has nothing to do synchronously, with no allocations, couldn't we use it in `submit` to avoid the expensive path if it's not required?

bhalevy replied:
By the way, @raphaelsc, I see that `strategy_control::candidates_as_runs` is currently used only by tests.

bhalevy replied:
I'll drop the last patch from this series so we can expedite the other patches.
So we can continue the discussion over a separate PR.

raphaelsc replied:
> By the way, @raphaelsc, I see that `strategy_control::candidates_as_runs` is currently used only by tests.

I dropped the ball on the PR that wires it into ICS :-( It's a important patch, I'll revive it today.

raphaelsc replied:
I think you're right @bhalevy. We can extract all the code for evaluation, including check if we can run the compaction (based on the weight of its descriptor), the registration of compacting, and forward all of that to regular_compaction_task.

Then regular_compaction_task, can use a do while {}. do is guaranteed to execute a task, whereas only next iterations will reevaluate again.

raphaelsc replied:
@bhalevy one last item that was in my mind, we probably still need to run the evaluation code under compaction sched group, to not interfere with foreground activity. but we can easily do it with a background fiber, e.g.:

`submit() { with_gate(...) { switch_to(compaction_sched); auto ret = run_reevaluation_code(); if (ret) { create_task(ret); } } }`

bhalevy replied:
> @bhalevy one last item that was in my mind, we probably still need to run the evaluation code under compaction sched group, to not interfere with foreground activity. but we can easily do it with a background fiber, e.g.:
> 
> `submit() { with_gate(...) { switch_to(compaction_sched); auto ret = run_reevaluation_code(); if (ret) { create_task(ret); } } }`

I'm not sure exactly where this would fit.

raphaelsc replied:
something like:

```
void compaction_manager::submit(table_state& t) {
    ...

    (void) with_gate(...) { 
        switch_to(compaction_sched);
        
        optional<struct whatever> ret = reevaluate_regular_compaction_for(t);
        
        if (ret) { 
            co_await perform_compaction<regular_compaction_task_executor>(..., std::move(ret));
        }
     }
}
```

bhalevy replied:
ok, that makes more sense now.

I just submitted v2, without the patch changing submit.
Let's work on that as a follow up.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5815/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5815/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7184/)
#### Build Details:
- Duration: 2 hr 17 min
- Builder: spider1.cloudius-systems.com

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -697,7 +697,9 @@ compaction_manager::compaction_reenabler::~compaction_reenabler() {
-            _cm.submit(*_table);
+            // We can safely call get_candidates here since
+            // compaction is known to be disabled.
+            _cm.submit(*_table, _cm.get_candidates(*_table));
```
Why not do this inside `submit()` improving it for all callers?

bhalevy replied:
I'm not sure calling get_candidates in all paths calling submit is safe.
Here, we know it's safe in the compaction reenabler since we're sure compaction is disabled.

bhalevy commented:
I also noticed that submit is called for all compaction groups in `table::trigger_compaction`.
And the latter is called from various places, including whenever sstables are added to the table.
So although we know the compaction group we added the sstable to, we trigger compaction on all compaction groups which is quite wasteful.

bhalevy commented:
In v2 ([0d937f3](https://github.com/scylladb/scylladb/commit/0d937f39746dc8da223bc5fcc6e2a4d418c8d1d6))
- rebased
- dropped "compaction_manager: compaction_reenabler: submit candidates for regular compaction"
- added "table: add_sstable_and_update_cache: trigger compaction only in compaction group"

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5836/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7202/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5836/)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: i-0c7784f5026eb47e3 (m5d.12xlarge)

bhalevy commented:
ci failed due to spot termination, rerunning

avikivity commented:
> This short series prevents the creation of compaction tasks when we know in advance that they have nothing to do. This is possible in the clean path by:
> 
> * improve the detection of candidates for cleanup by skipping sstables that require cleanup but are already being compacted
> * checking that list of sstables selected for cleanup isn't empty before creating the cleanup task
> 
> For upgrade sstables, and generally when rewriting all sstable: launch the task only if the list off candidate sstables isn't empty.
> 
> For regular compaction, when triggered via `table::add_sstable_and_update_cache`, we currently trigger compaction (by calling `submit`) on all compaction groups while the sstable is added only to one of them.

We should call compaction_group::add_sstable(), which should call its table to update the cache.

Notifying 100 compaction groups each time a compaction finishes will take its toll. That's a 10,000X amplification, since compactions are 100X more frequent with compaction groups.

> Also, it is typically called for maintenance sstables that are awaiting offstrategy compaction, in which case we can skip calling `submit` entirely since the caller triggers offstrategy compaction at a later stage.
> 
> Refs #15673 Refs #16694 Fixes #16803

raphaelsc commented:
On Wed, Jan 17, 2024 at 3:18 PM Avi Kivity ***@***.***> wrote:

> This short series prevents the creation of compaction tasks when we know
> in advance that they have nothing to do. This is possible in the clean path
> by:
>
>    - improve the detection of candidates for cleanup by skipping sstables
>    that require cleanup but are already being compacted
>    - checking that list of sstables selected for cleanup isn't empty
>    before creating the cleanup task
>
> For upgrade sstables, and generally when rewriting all sstable: launch the
> task only if the list off candidate sstables isn't empty.
>
> For regular compaction, when triggered via
> table::add_sstable_and_update_cache, we currently trigger compaction (by
> calling submit) on all compaction groups while the sstable is added only
> to one of them.
>
> We should call compaction_group::add_sstable(), which should call its
> table to update the cache.
>
> Notifying 100 compaction groups each time a compaction finishes will take
> its toll. That's a 10,000X amplification, since compactions are 100X more
> frequent with compaction groups.
>

This was only a problem when adding sstables from an external agent, e.g.
refresh or table population. Compaction and flush are already handled
properly by only notifying the underlying group.


> Also, it is typically called for maintenance sstables that are awaiting
> offstrategy compaction, in which case we can skip calling submit entirely
> since the caller triggers offstrategy compaction at a later stage.
>
> Refs #15673 <https://github.com/scylladb/scylladb/issues/15673> Refs
> #16694 <https://github.com/scylladb/scylladb/issues/16694> Fixes #16803
> <https://github.com/scylladb/scylladb/issues/16803>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/16808#issuecomment-1896362910>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKYA42IK4SJ2BJFX63S2VLYPAIWJAVCNFSM6AAAAABB5PC6XWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQOJWGM3DEOJRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5901/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7241/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5901/)
#### Build Details:
- Duration: 1 hr 46 min
- Builder: spider8.cloudius-systems.com

avikivity commented:
> > This short series prevents the creation of compaction tasks when we know in advance that they have nothing to do. This is possible in the clean path by:
> > 
> > * improve the detection of candidates for cleanup by skipping sstables that require cleanup but are already being compacted
> > * checking that list of sstables selected for cleanup isn't empty before creating the cleanup task
> > 
> > For upgrade sstables, and generally when rewriting all sstable: launch the task only if the list off candidate sstables isn't empty.
> > For regular compaction, when triggered via `table::add_sstable_and_update_cache`, we currently trigger compaction (by calling `submit`) on all compaction groups while the sstable is added only to one of them.
> 
> We should call compaction_group::add_sstable(), which should call its table to update the cache.
> 
> Notifying 100 compaction groups each time a compaction finishes will take its toll. That's a 10,000X amplification, since compactions are 100X more frequent with compaction groups.
> 
> > Also, it is typically called for maintenance sstables that are awaiting offstrategy compaction, in which case we can skip calling `submit` entirely since the caller triggers offstrategy compaction at a later stage.
> > Refs #15673 Refs #16694 Fixes #16803

@xemul Im reminded of the problem we see with 100 tablets/shard. Maybe every compaction group is updating the cache, even though the sstable only belongs to one of them.

avikivity commented on replica/table.cc:
```diff
@@ -826,20 +826,24 @@ table::do_add_sstable_and_update_cache(sstables::shared_sstable sst, sstables::o
-    co_await do_add_sstable_and_update_cache(std::move(sst), offstrategy);
-    trigger_compaction();
+    bool do_trigger_compaction = offstrategy == sstables::offstrategy::no;
+    co_await do_add_sstable_and_update_cache(std::move(sst), offstrategy, do_trigger_compaction);
```
Maybe we should rename this to import_sstable. The normal compaction behavior is handled in compaction_group IIUC.

can offstrategy be no here?

bhalevy replied:
When called from `table_populator::populate_subdir` is may be "no" under some conditions
And with https://github.com/scylladb/scylladb/pull/16863 it might also be "no" for repair with TWCS.
